### PR TITLE
TPS refactoring: support scraping from genesis, or last finalised + parablock tps measurements

### PR DESCRIPTION
**Pull request by bernardoaraujor**
Versi needs the `tps` utility to start scraping from the latest finalized block, instead of genesis.

This PR adds a new boolean CLI arg. If `tps` is called with `--genesis`, then it starts scraping from genesis.
If nothing is provided, it starts scraping from the latest finalized block.

---

The PR is still a draft. We should use this to also refactor the way `tps` is outputting the metrics, so that [printing on console](https://github.com/paritytech/polkadot-stps/blob/d533fdc589705ea718abb79a8c8b03bed5850695/utils/tps/src/main.rs#L64) is not the only option.

@bredamatt what would be the best output for the metrics?

---

Edit by @bredamatt :

I have added a few additional parameters;

- whereas `genesis` is useful for separating between zombienet and Versi based deployments when tracking TPS on the relay chain, we also need to be more explicit as the method to track TPS on the relay chain and para chain side differs.
- therefore, I added a `para_finality` boolean argument which, when set to `true`, makes `tps` a concurrent service, and spawns both a collator RPC client and a relay chain RPC client concurrently. The relay chain client subscribes to `CandidateIncluded` events on the relay chain side, and fetches the `ParaBlock` head (the hash of the parablock) once the parablock has passed the inclusion (consensus) pipeline relay chain side. It then sends this hash to the collator RPC client via an async mpsc channel. 
- The collator RPC client can thereafter lookup details about the parablock given its hash, including the number of `Transfer` events that has occurred in it. 
- Because the time-difference between the previous parablock hash and the most recent parablock hash needs to be calculated, I have accounted for errors in fetching the previous block by assuming a sync-backing default parablock time of 12 seconds to start with by passing in `default_parablock_time` argument with a default setting. When async-backing comes, this default time should be set to 6 seconds. It is expected that we at least double TPS once async-backing comes accordingly as the time-difference will be halved, ceteris paribus.

A quick note regarding `subxt` metadata: 
Because both the `polkadot-parachain` / `tick-collator` and the `rococo` metadata include the `Balances` pallet, then it is possible to have the same runtime types for both the collator and relay chain RPC clients. This would not be the case if we were to test different pallets which are not included in both runtimes accordingly.

### CODE REVIEW

**Comment by bernardoaraujor** [comment]
cool, I'll look into it and try to write something by EOW

**Comment by bredamatt** [comment]
I committed the changes related to this issue: https://github.com/paritytech/polkadot-stps/issues/45 on this branch. LMK what you think @bernardoaraujor

File: tests/native/para-native.json [hunk 1]
```diff
@@ -22,15 +22,6 @@
   },
   "parachains": [
     {
-      "genesis": {
```
Comment (bernardoaraujor) [inline tests/native/para-native.json:25]: where will the prefunded accounts come from?

File: tests/native/para-native.json [hunk 2]
```diff
@@ -22,15 +22,6 @@
...
-        "runtime": {
-          "runtime_genesis_config": {
-            "balances": {
-              "balances": {% include "../funded-accounts.json" %}
-            }
-          }
-        }
-      },
```
Comment (bernardoaraujor) [inline tests/native/para-native.json:33]: where will the pre-funded accounts come from?

File: utils/tps/src/main.rs [hunk 3]
```diff
@@ -44,7 +161,7 @@ pub async fn calc_tps(api: &Api, n: usize, genesis: bool) -> Result<(), Error> {
 				.unwrap();
 			(2, block_timestamp_genesis)
 		},
-		false => {
+		true => {
```
Comment (bernardoaraujor) [inline utils/tps/src/main.rs:169]: why were these booleans flipped?

the original rationale was: if `--genesis` is present as a CLI arg (which implies `genesis == true`), when we start scraping from genesis

now, we're explicitly calling `--genesis` when we **don't** want to scrape from genesis?

**Comment by bernardoaraujor** [comment]
why are we adding a new `parachain-monitor` crate, and then replicating the same code inside `tps`?
do we have any specific reason to have a new separate crate?

adding the same code twice in different places feels reduntant.

File: tests/native/para-native.json (hunk 2 above)
Comment (bernardoaraujor) [reply tests/native/para-native.json:33]: oh I guess Versi does that in a different way? I had forgotten about that for a second

anyways, how do we make sure `sender` is using the right accounts?

**Comment by bredamatt** [comment]
As mentioned we can delete the Parachain monitor crate once fully integrated, which it is now.

File: tests/native/para-native.json (hunk 2 above)
Comment (bredamatt) [reply tests/native/para-native.json:33]: Versi does it differently indeed - the chain spec is basically already populated with the pre funded accounts. 

Regarding the zombienet side, let's add the configuration necessary once it's supported in zombienet. The previous configuration kept throwing "Bad configuration", so I removed it.

File: utils/tps/src/main.rs (hunk 3 above)
Comment (bredamatt) [reply utils/tps/src/main.rs]: Yep typo, my bad! Fixed it back now.

**Review by bredamatt — APPROVED** [review]

File: utils/tps/src/main.rs [hunk 4]
```diff
@@ -18,39 +37,178 @@ struct Args {
...
+			},
+			// Assume default if unable to get the previous parablock from parablock number
+			None => {
+				warn!(
+					"TPS Counter ===> Assuming default parablock time of: {:?}",
+					default_parablock_time
+				);
+				Duration::as_secs_f64(&Duration::new(default_parablock_time, 0)) as u64
+			},
+		};
```
Comment (bredamatt) [inline utils/tps/src/main.rs:100]: The `unwrap()`s here have to be fixed - just noticed an error during testing.

Comment (bernardoaraujor) [reply utils/tps/src/main.rs:100]: what was the error?

File: utils/tps/src/main.rs [hunk 5]
```diff
@@ -18,39 +37,178 @@ struct Args {
 	/// Total number of expected transactions
 	#[arg(long)]
 	num: usize,
+
+	/// Whether to subscribe to blocks from genesis or not.
+	/// For zombienet tests, this should be set to true.
+	/// When deploying tps in more long-living networks, set this to false.
```
Comment (bernardoaraujor) [inline utils/tps/src/main.rs:43]:
```suggestion
	/// When deploying tps in more long-living networks, set this to false (or simply omit it).
```

File: utils/tps/src/main.rs [hunk 6]
```diff
@@ -18,39 +37,184 @@ struct Args {
...
+			}
+		}
+		// Print to stdout and reset counter
+		info!(
+			"TPS Counter ===> Counted {} TPS in ParaHead: {:?}",
+			trx_in_parablock / parablock_time,
+			para_head
+		);
+	}
+	trx_in_parablock = 0;
```
Comment (bernardoaraujor) [inline utils/tps/src/main.rs:153]: shouldn't this line be inside the `while` loop?

Comment (bernardoaraujor) [inline utils/tps/src/main.rs:153]: shouldn't we be resetting `trx_in_parablock` inside the `while` loop?

**Comment by bernardoaraujor** [comment]
We can keep working on this PR, but https://github.com/paritytech/polkadot-stps/pull/48 should be merged before we merge this.

File: utils/tps/src/main.rs (hunk 4 above)
Comment (bredamatt) [reply utils/tps/src/main.rs:131]: Panic due to unwrap on `None` value.

File: utils/tps/src/main.rs (hunk 6 above)
Comment (bredamatt) [reply utils/tps/src/main.rs:150]: Yes I believe we should, indeed.

**Comment by bredamatt** [comment]
https://github.com/paritytech/polkadot-stps/pull/48 Approved on my end. We can always improve the metrics further down the line. 

For now what is critical is having the TPS metrics.

File: utils/tps/src/main.rs (hunk 4 above)
Comment (bernardoaraujor) [reply utils/tps/src/main.rs:131]: I believe you actually need something like this:
```
				let parablock_timestamp = para_api
					.storage()
-					.fetch(&storage_timestamp_storage_addr, Some(para_head))
+					.fetch(&storage_timestamp_storage_addr, Some(hash))
					.await?
					.unwrap();
				let previous_parablock_timestamp = para_api
					.storage()
-					.fetch(&storage_timestamp_storage_addr, Some(hash))
+					.fetch(&storage_timestamp_storage_addr, Some(previous_hash))
					.await?
					.unwrap();
```

Comment (bernardoaraujor) [reply utils/tps/src/main.rs]: fixed here: https://github.com/paritytech/polkadot-stps/pull/43/commits/74dca75fe69417cf8d3730271f5112add58cae1b

**Pull request merged**